code/datums/status_effects/debuffs/fire_stacks.dm

```diff
@@ -0,0 +1,373 @@
+	var/list/merge_types = list()
+	/// What status effect types do we override if they exist. These are simply deleted when forced.
+	var/list/override_types = list()
+	/// For how much firestacks does one our stack count
+	var/stack_modifier = 1
+
+/datum/status_effect/fire_handler/refresh(mob/living/new_owner, new_stacks, forced = FALSE)
+	if(forced)
+		set_stacks(new_stacks)
+	else
+		adjust_stacks(new_stacks)
+
+/datum/status_effect/fire_handler/on_creation(mob/living/new_owner, new_stacks, forced = FALSE)
+	if(isanimal(owner))
+		qdel(src)
+		return
+
+	owner = new_owner
+	set_stacks(new_stacks)
+
+	for(var/enemy_type in enemy_types)
+		var/datum/status_effect/fire_handler/enemy_effect = owner.has_status_effect(enemy_type)
+		if(enemy_effect)
+			if(forced)
+				qdel(enemy_effect)
+				continue
+
+			var/cur_stacks = stacks
+			adjust_stacks(-enemy_effect.stacks * enemy_effect.stack_modifier / stack_modifier)
+			enemy_effect.adjust_stacks(-cur_stacks * stack_modifier / enemy_effect.stack_modifier)
+			if(enemy_effect.stacks <= 0)
+				qdel(enemy_effect)
+
+			if(stacks <= 0)
+				qdel(src)
+				return
+
+	if(!forced)
+		var/list/merge_effects = list()
+		for(var/merge_type in merge_types)
+			var/datum/status_effect/fire_handler/merge_effect = owner.has_status_effect(merge_type)
+			if(merge_effect)
+				merge_effects += merge_effects
+
+		if(LAZYLEN(merge_effects))
+			for(var/datum/status_effect/fire_handler/merge_effect in merge_effects)
+				merge_effect.adjust_stacks(stacks * stack_modifier / merge_effect.stack_modifier / LAZYLEN(merge_effects))
+			qdel(src)
+
+	for(var/override_type in override_types)
+		var/datum/status_effect/fire_handler/override_effect = owner.has_status_effect(override_type)
+		if(override_effect)
+			if(forced)
+				qdel(override_effect)
+				continue
+
+			adjust_stacks(override_effect.stacks)
+			qdel(override_effect)
+
+	. = ..()
+	cache_stacks()
+
+/**
+ * Setter and adjuster procs for firestacks
+ *
+ * Arguments:
+ * - new_stacks
+ *
+ */
+
+/datum/status_effect/fire_handler/proc/set_stacks(new_stacks)
+	stacks = max(0, min(stack_limit, new_stacks))
+	cache_stacks()
+
+/datum/status_effect/fire_handler/proc/adjust_stacks(new_stacks)
+	stacks = max(0, min(stack_limit, stacks + new_stacks))
+	cache_stacks()
+
+/**
+ * Refresher for mob's fire_stacks
+ */
+
+/datum/status_effect/fire_handler/proc/cache_stacks()
+	owner.fire_stacks = 0
+	owner.on_fire = FALSE
+	owner.clear_alert(ALERT_FIRE)
+	for(var/fire_type in subtypesof(/datum/status_effect/fire_handler))
+		var/datum/status_effect/fire_handler/possible_fire = owner.has_status_effect(fire_type)
+		if(possible_fire)
+			owner.fire_stacks += possible_fire.stacks * possible_fire.stack_modifier
+			if(istype(possible_fire))
+				var/datum/status_effect/fire_handler/fire_stacks/our_fire = possible_fire
+				if(our_fire.on_fire)
+					owner.on_fire = TRUE
+					owner.throw_alert(ALERT_FIRE, our_fire.fire_alert_type)
```

Comment:
Rather then clearing the alert and on fire, then resetting it, it'd be better to have a var that's called "should_fire" or something. Defaults to FALSE, if it finds a source of fire, set it to true
Then use the owner's old on_fire var to decide it we should clear the alert, send a new one, or do nothing.